### DISCUSSION TITLE
Emgu.CV with Avalonia.Image

### DISCUSSION THREAD

**Discussion by vo1s** [AvaloniaUI/Avalonia#13748] [Q&A] [answered]
I am trying to get video stream from camera using Emgu.CV and RTSP protocol and display it in avalonia app. I have this class:
```
public class VideoStreamModel
{
    private VideoCapture capture;

    public VideoStreamModel(string rtspUrl)
    {
        capture = new VideoCapture(rtspUrl);
    }

    public Mat GetNextFrame()
    {
        return capture.QueryFrame();
    }

    public void Dispose()
    {
        capture?.Dispose();
    }
}
```
I use it and MainViewModel.cs:
```
public partial class MainViewModel : ObservableObject
{
    private readonly DbscanVladContext _dbMain;
    VideoStreamModel model = new VideoStreamModel("rtsp://admin:admin@192.168.3.103:554/av0_0");
    [ObservableProperty] private Image _videoFrame = new Image();
....
```
And just for test in constructor getting frames and set it to the image in xaml file:
```
public MainViewModel()
    {
        _dbMain = new DbscanVladContext();
        DispatcherTimer timer = new DispatcherTimer();
        timer.Interval = TimeSpan.FromMilliseconds(33);
        timer.Tick += cameraFPS;
        timer.Start();
        
    }

    private void cameraFPS(object sender, EventArgs e)
    {
        Mat frame = model.GetNextFrame();

        if (frame != null)
        {
            var bitmapp = frame.ToBitmap().ConvertToAvaloniaBitmap();
            Image bitmapSource = new Image { Source = bitmapp };
            VideoFrame = bitmapSource;
        }
```
`ConvertToAvaloniaBitmap()` is a method found here: [ConvertToAvaloniaBitmap()](https://github.com/AvaloniaUI/Avalonia/discussions/5908)
MainWindow.xaml:
```
<StackPanel  Width="650" Height="650">
                        <Image Source="{Binding VideoFrame, Mode=TwoWay}" Width="660" Height="700"/>
                    </StackPanel>
```
So the problem is that I dont get any image displayed when I run it. The connection is working and I get pictures saved to file when I was testing it in console app.

**Comment by timunie** [comment]
`            Image bitmapSource = new Image { Source = bitmapp };` can't you just return the Bitmap instead of a control?

**Reply by vo1s** [reply]
didn't actually get your point here. Please show me.

**Reply by timunie** [reply]
Image is a Control. Your ViewModel should not know any controls.

**Reply by vo1s** [reply]
so it doesn't work like this neither 
```
var bitmapp = frame.ToBitmap().ConvertToAvaloniaBitmap();
VideoFrame = new Image { Source = bitmapp };
```
Do you mean I need to bind to the bitmap instead?

**Comment by thevortexcloud** [comment]
You are not the first person to try to show a camera feed in Avalonia. From memory writable bitmap is better here for this since it lets you reuse the instance. But there may be better options, it's not something I have tried.

**Comment by vo1s** [comment, marked as answer]
I've made VideoFrame as a Bitmap and bind directly to it.
```
[ObservableProperty] private Bitmap _videoFrame;
```
```
var bitmapp = frame.ToBitmap().ConvertToAvaloniaBitmap();
            VideoFrame = bitmapp;
```
